File: SatinalmaPro.Shared/Helpers/BildirimMantikAnahtari.cs
using System.Security.Cryptography;
using System.Text;
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Helpers;

/// <summary>Aynı iş akışı bildirimini tekilleştirmek için mantıksal anahtar (talep + tip + hedef).</summary>
public static class BildirimMantikAnahtari
{
    public static string Olustur(BildirimKaydi bildirim)
    {
        // Inbox kaynaklı / talepsiz kayıtlar: docId ile sabitle — her yüklemede Guid.NewGuid()
        // ile şişen veri/bildirimler belgesini önler.
        if (!string.IsNullOrWhiteSpace(bildirim.InboxDocId))
            return $"inbox:{bildirim.InboxDocId.Trim()}";

        if (bildirim.TalepId is not { } talepId)
            return $"id:{bildirim.Id}";

        var hedef = !string.IsNullOrWhiteSpace(bildirim.HedefUid)
            ? $"u:{bildirim.HedefUid.Trim()}"
            : $"r:{(bildirim.HedefRol ?? "").Trim()}";

        return $"{talepId:N}:{bildirim.Tip}:{hedef}";
    }

    /// <summary>Firestore inbox documentId → kararlı Guid (her okumada yeni id üretmez).</summary>
    public static Guid InboxDocIddenGuid(string docId)
    {
        if (Guid.TryParse(docId, out var g))
            return g;

        var bytes = MD5.HashData(Encoding.UTF8.GetBytes("inbox:" + docId.Trim()));
        return new Guid(bytes);
    }
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/android/core/helpers/BildirimMantikAnahtari.kt
package com.satinalmapro.android.core.helpers

import com.satinalmapro.android.core.model.BildirimRecord

/** Aynı iş akışı bildirimini tekilleştirmek için mantıksal anahtar (talep + tip + hedef). */
object BildirimMantikAnahtari {
    fun olustur(bildirim: BildirimRecord): String {
        val talepId = normalizeTalepId(bildirim.talepId)
        if (talepId.isNullOrBlank()) return "id:${bildirim.id}"

        val hedef = if (!bildirim.hedefUid.isNullOrBlank()) {
            "u:${bildirim.hedefUid.trim().lowercase()}"
        } else {
            "r:${bildirim.hedefRol.orEmpty().trim()}"
        }
        return "$talepId:${bildirim.tip}:$hedef"
    }

    /** Masaüstü ile uyumlu GUID normalizasyonu ({talepId:N}). */
    fun normalizeTalepId(talepId: String?): String? =
        talepId?.trim()?.replace("-", "")?.lowercase()?.takeIf { it.isNotBlank() }
}
